File: practice/algorithms/sorting/almost_sorted/almost_sorted.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

int main()
{
    int n;
    cin >> n;
    
    vector<int> a(n);
    for (int i = 0; i < n; i++)
    {
        cin >> a[i];
    }
    
    if (is_sorted(a.begin(), a.end()))
    {
        cout << "yes" << endl;
    }
    else
    {
        int begin, end;
        
        begin = 0;
        while (begin < n - 1 && a[begin] < a[begin + 1])
        {
            begin++;
        }
        
        end = n - 1;
        while (end > 0 && a[end] > a[end - 1])
        {
            end--;
        }
        
        swap(a[begin], a[end]);
        if (is_sorted(a.begin(), a.end()))
        {
            cout << "yes" << endl;
            cout << "swap " << (begin + 1) << " " << (end + 1) << endl;
        }
        else
        {
            swap(a[begin], a[end]);
            reverse(a.begin() + begin, a.begin() + end + 1);
            if (is_sorted(a.begin(), a.end()))
            {
                cout << "yes" << endl;
                cout << "reverse " << (begin + 1) << " " << (end + 1) << endl;
            }
            else
            {
                cout << "no" << endl;
            }
        }
    }
    
    return 0;
}

File: practice/algorithms/sorting/almost_sorted/almost_sorted.hs
import Data.List

isSorted :: [Int] -> Bool
isSorted [] = True
isSorted [x] = True
isSorted a = and $ zipWith (<=) a (tail a)

findBegin :: [Int] -> Int
findBegin [] = 0
findBegin [x] = 0
findBegin (x:y:a)
    | x <= y = 1 + findBegin (y:a)
    | otherwise = 0

findEnd :: [Int] -> Int
findEnd a = length a - 1 - findEnd' (reverse a) where
    findEnd' [] = 0
    findEnd' [x] = 0
    findEnd' (x:y:a)
        | x >= y = 1 + findEnd' (y:a)
        | otherwise = 0

swapElementsAt :: Int -> Int -> [Int] -> [Int]
swapElementsAt a b list = list1 ++ [list !! b] ++ list2 ++ [list !! a] ++ list3
    where   list1 = take a list
            list2 = drop (succ a) (take b list)
            list3 = drop (succ b) list
            
reverseSegment a b list = list1 ++ list2 ++ list3
    where   list1 = take a list
            list2 = reverse $ take (b - a + 1) (drop a list)
            list3 = drop (succ b) list

isAlmostSorted a
    | isSorted a = "yes"
    | isSorted s = "yes\nswap " ++ (show (begin + 1)) ++ " " ++ (show (end + 1))
    | isSorted t = "yes\nreverse " ++ (show (begin + 1)) ++ " " ++ (show (end + 1))
    | otherwise  = "no" where
        begin = findBegin a
        end   = findEnd a
        s = swapElementsAt begin end a
        t = reverseSegment begin end a
        
main = do
    n <- readLn :: IO Int
    ar <- getLine
    let a = map (read :: String -> Int) $ words ar
    putStrLn $ isAlmostSorted a
